lib/elixir/lib/kernel/typespec.ex

```diff
@@ -535,10 +535,29 @@ defmodule Kernel.Typespec do
   end
 
   # Handle type operator
-  defp typespec({:::, meta, [var, expr]}, vars, caller) do
-    left = typespec(var, [elem(var, 0) | vars], caller)
-    right = typespec(expr, vars, caller)
-    {:ann_type, line(meta), [left, right]}
+  defp typespec({:::, meta, [var, expr]} = orig, vars, caller) do
+    case var do
+      {var_name, _meta, context} when is_atom(var_name) and is_atom(context) ->
```

Comment:
Maybe we can move this matching to the function head and split the current code in two clauses? Other than that, everything looks good!